test/distributed/fsdp/test_fsdp_optim_state.py

```diff
@@ -1439,6 +1440,70 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         loss.backward()
         optim.step()
 
+    @skip_if_lt_x_gpu(2)
+    def test_compatible_with_named_optimizer(self):
+        class TestDummyModel(torch.nn.Module):
+            def __init__(self):
+                super(TestDummyModel, self).__init__()
+                torch.manual_seed(0)
+                self.net1 = nn.Sequential(nn.Linear(8, 16), nn.ReLU())
+                self.net2 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+                self.net3 = nn.Linear(32, 64)
+                self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+            def forward(self, x):
+                return self.net4(self.net3(self.net2(self.net1(x))))
+
+        models = []
+        optims = []
+        state_dicts = []
+        models.append(FSDP(TestDummyModel().cuda(), use_orig_params=True))
+        optims.append(torch.optim.Adam(models[-1].parameters(), lr=1e-2))
+        models.append(FSDP(TestDummyModel().cuda(), use_orig_params=True))
+        optims.append(
+            _NamedOptimizer(
+                models[-1].named_parameters(),
+                torch.optim.Adam,
+                [{"params": models[-1].parameters()}],
+                lr=1e-2,
+            )
+        )
+
+        # Train one batch and see if optim_state_dict are the same.
+        batch = torch.rand(5, 8)
+        for model, optim in zip(models, optims):
+            loss = model(batch).sum()
+            loss.backward()
+            optim.step()
+            state_dicts.append(FSDP._optim_state_dict(model, optim))
+
+        self._check_same_param_groups(
+            state_dicts[0], state_dicts[1], check_same_param_keys=False
+        )
+        self._check_same_state(
+            state_dicts[0], state_dicts[1], check_same_param_keys=True
+        )
+
+        # Make optim1 has a different state.
+        for i in range(5):
+            batch = torch.rand(5, 8)
+            loss = models[1](batch).sum()
+            loss.backward()
+            optims[1].step()
+
+        # Load the state back to see if load_optim_state_dict works.
+        optims[1].load_state_dict(
+            FSDP._load_optim_state_dict_pre_hook(models[1], optims[1], state_dicts[1])
+        )
+        state_dicts[1] = FSDP._optim_state_dict(models[1], optims[1])
+
+        self._check_same_param_groups(
+            state_dicts[0], state_dicts[1], check_same_param_keys=False
+        )
+        self._check_same_state(
```

Comment:
No. Only `optim1` steps 5 times so there will be difference. However, we then load the state_dict back so that we should get the same state. This logic is to test the correctness of loading.